File: app/routers/websocket.py
"""
WebSocket endpoint for real-time job progress updates.
"""

from fastapi import APIRouter, WebSocket, WebSocketDisconnect
from ..services.websocket_manager import manager
from ..services.job_manager import get_job
from ..models import JobStatus

router = APIRouter(tags=["websocket"])


@router.websocket("/ws/jobs/{job_id}")
async def websocket_job_status(websocket: WebSocket, job_id: str):
    """
    WebSocket endpoint for real-time job status updates.

    Connect to this endpoint after creating a job to receive
    real-time progress updates instead of polling.

    Messages sent to client:
    {
        "type": "job_update",
        "job_id": "...",
        "status": "pending|processing|completed|failed",
        "progress": 0-100,
        "message": "...",  # optional
        "error": "...",    # optional, on failure
        "download_url": "..."  # optional, on completion
    }
    """
    # Must accept websocket before we can close it or send messages
    await websocket.accept()

    # Check if job exists
    job = get_job(job_id)
    if not job:
        await websocket.close(code=4004, reason="Job not found")
        return

    # Register this connection to receive broadcasts
    await manager.connect(websocket, job_id)

    try:
        # Send current job status immediately
        initial_status = {
            "type": "job_update",
            "job_id": job_id,
            "status": job["status"],
            "progress": job.get("progress", 0),
        }
        if job["status"] == JobStatus.COMPLETED:
            initial_status["download_url"] = f"/api/posters/{job_id}"
        if job.get("error"):
            initial_status["error"] = job["error"]

        await websocket.send_json(initial_status)

        # Keep connection alive and wait for messages (heartbeat)
        while True:
            # Wait for any message from client (ping/pong or close)
            data = await websocket.receive_text()
            # Client can send "ping" to keep connection alive
            if data == "ping":
                await websocket.send_json({"type": "pong"})

    except WebSocketDisconnect:
        pass
    finally:
        await manager.disconnect(websocket, job_id)

File: server/src/routes/websocket.js
import { getJob, setNotifyCallback } from '../services/jobManager.js';

// Track active WebSocket connections by job ID
const connections = new Map();

/**
 * Notify all connected clients about a job update.
 * @param {string} jobId - The job ID
 * @param {Object} update - The update data
 */
function notifyJobUpdate(jobId, update) {
  const jobConnections = connections.get(jobId);
  if (!jobConnections || jobConnections.size === 0) {
    return;
  }

  const message = JSON.stringify({
    type: 'job_update',
    ...update,
  });

  for (const ws of jobConnections) {
    if (ws.readyState === 1) { // OPEN
      ws.send(message);
    }
  }
}

// Set the notification callback for the job manager
setNotifyCallback(notifyJobUpdate);

/**
 * Set up WebSocket routes on the app.
 * @param {Express.Application} app - The Express app with express-ws
 */
export function setupWebSocket(app) {
  /**
   * WebSocket endpoint for job updates.
   * /ws/jobs/:jobId
   */
  app.ws('/ws/jobs/:jobId', (ws, req) => {
    const { jobId } = req.params;
    console.log(`[WebSocket] Client connected for job ${jobId}`);

    // Store connection
    if (!connections.has(jobId)) {
      connections.set(jobId, new Set());
    }
    connections.get(jobId).add(ws);

    // Heartbeat to keep connection alive (every 30 seconds)
    const heartbeat = setInterval(() => {
      if (ws.readyState === 1) { // OPEN
        ws.send(JSON.stringify({ type: 'ping' }));
      }
    }, 30000);

    // Send current job status immediately
    const job = getJob(jobId);
    if (job) {
      ws.send(JSON.stringify({
        type: 'job_update',
        job_id: jobId,
        status: job.status,
        progress: job.progress,
        message: job.message,
        error: job.error,
        download_url: job.status === 'completed' ? `/api/posters/${jobId}` : null,
      }));
    }

    // Handle client disconnect
    ws.on('close', () => {
      console.log(`[WebSocket] Client disconnected for job ${jobId}`);
      clearInterval(heartbeat);
      const jobConnections = connections.get(jobId);
      if (jobConnections) {
        jobConnections.delete(ws);
        if (jobConnections.size === 0) {
          connections.delete(jobId);
        }
      }
    });

    // Handle errors
    ws.on('error', (error) => {
      console.error(`[WebSocket] Error for job ${jobId}:`, error);
    });
  });
}
